src/core/layers/layer.py

```diff
@@ -59,13 +60,205 @@ def _hideWidgetFid(self, layer: QgsVectorLayer) -> None:
+            print(type(e).__name__ + ": " + str(e))
+
+    def stop(self):
+        if self.playing:
+            self.stream.stop()
+            self.playing = False
+
+
+class AudioPlayerDialog(QDialog):
+    def __init__(self, filename):
+        super().__init__()
+        self.player: AudioPlayer
+        self.thread: AudioThread
+        self.filename = filename
+        self.setWindowTitle(Path(self.filename).name)
+
+        self.playButton = QPushButton()
+        self.playButton.setCheckable(True)
+        self.playButton.clicked.connect(self.toggleSound)
+        self.playing = False
+
+        self.slider = QSlider(Qt.Horizontal)
+        self.slider.setMinimum(0)
+        sound, fs = sf.read(filename)
+        self.slider.setMaximum(int(sound.shape[0] / fs))
+        self.slider.setSingleStep(1)
+
+        self.label = QLabel(
+            timeformat(0) + " / " + timeformat(self.slider.maximum())
+        )
+        self.slider.valueChanged.connect(self.updateLabel)
+
+        self.Vlayout = QVBoxLayout(self)
+        self.Hlayout = QHBoxLayout()
+        self.Hlayout.addWidget(self.playButton)
+        self.Hlayout.addWidget(self.label)
+        self.Vlayout.addLayout(self.Hlayout)
+        self.Vlayout.addWidget(self.slider)
+
+    @property
+    def playing(self):
+        return self._playing
+
+    @playing.setter
+    def playing(self, state: bool):
+        if state:
+            self.playButton.setIcon(
+                QIcon(\"{
+                    (
+                        Path(__file__).parent.parent.parent.parent
+                        / "images"
+                        / "pause.png"
+                    ).as_posix()
+                }\")
+            )
+        else:
+            self.playButton.setIcon(
+                QIcon(\"{
+                    (
+                        Path(__file__).parent.parent.parent.parent
+                        / "images"
+                        / "play.png"
+                    ).as_posix()
+                }\")
+            )
+        self._playing = state
+
+    def toggleSound(self):
+        if self.playButton.isChecked():
+            self.thread = AudioThread(
+                self.filename, self.slider.sliderPosition()
+            )
+            self.thread.time.connect(self.updateSlider)
+            self.thread.finished.connect(self.end)
+            self.thread.start()
+            self.playing = True
+        else:
+            self.thread.quit()
+            self.playing = False
+
+    def updateSlider(self, value: int):
+        self.slider.setSliderPosition(value)
+
+    def updateLabel(self, value: int):
+        self.label.setText(
+            timeformat(value) + " / " + timeformat(self.slider.maximum())
+        )
+
+    def end(self):
+        self.thread.time.disconnect(self.updateSlider)
+        self.thread.quit()
+        self.playButton.setChecked(False)
+        self.playing = False
+        self.slider.setSliderPosition(0)
+
```

Comment:
Considering that the pyscript is pretty long now, could we move it in a dedicated file?

Reply:
it's done ! ;)